colossalai/pipeline/schedule/zero_bubble_pp.py

```diff
@@ -0,0 +1,848 @@
+                tensor=output_obj,
+                grad=output_obj_grad,
+                inputs=list(model_chunk[model_chunk_id].parameters()),
+                retain_graph=False,
+            )
+
+        else:
+            if self.stage_manager.is_first_stage(ignore_chunk=True):
+                optimizer.backward_by_grad(
+                    tensor=output_obj,
+                    grad=None,
+                    inputs=list(model_chunk[model_chunk_id].parameters()),
+                    retain_graph=False,
+                )
+            else:
+                optimizer.backward_by_grad(
+                    tensor=output_obj,
+                    grad=output_obj_grad,
+                    inputs=list(model_chunk[model_chunk_id].parameters()),
+                    retain_graph=False,
+                )
+
+    def schedule_f(
+        self,
+        scheduled_node,
+        model_chunk: torch.nn.ModuleList,
+        model_chunk_id: int,
+        criterion: Callable,
+        accum_loss: Optional[torch.Tensor] = None,
+        outputs: Optional[List[Any]] = None,
+    ):
+        """A complete forward schedule; Include recv fwd --> cal fwd --> send fwd;
+
+        Args:
+            scheduled_node:
+            model_chunk (ModuleList or Module): Model Chunk to be run;
+            model_chunk_id (int): The current model chunk idx;
+            criterion (Callable): loss function;
+            accum_loss (Optional[torch.Tensor], optional): Accumulated loss. Defaults to None.
+            outputs (Optional[List[Any]], optional): List to store the output of the last stage (final output). Defaults to None.
+
+        Returns:
+            Nothing.
+        """
+        micro_batch = self.load_micro_batch(model_chunk_id=model_chunk_id)
+        # Step1: recv fwd
+        if model_chunk_id == 0:
+            # is first stage; get input from func param
+            if self.stage_manager.is_first_stage(ignore_chunk=True):
+                input_obj = micro_batch
+            else:
+                input_obj = self.recv_forward_buffer[model_chunk_id].pop(0)
+        else:
+            # is last stage; recv from local
+            if self.stage_manager.is_last_stage(ignore_chunk=True):
+                input_obj = self.local_send_forward_buffer.pop(0)
+            # not last stage; recv from next
+            else:
+                input_obj = self.recv_forward_buffer[model_chunk_id].pop(0)
+
+        # Step2: fwd step
+        output_obj = self.forward_step(
+            model_chunk=model_chunk,
+            model_chunk_id=model_chunk_id,
+            input_obj=input_obj,
+            criterion=criterion,
+            accum_loss=accum_loss,
+            outputs=outputs,
+        )
+        # add input and output object for backward b
+        self.input_tensors[model_chunk_id].append(input_obj)
+
+        # detached output; for bwd b&w, we only need the graph(grad_fn) of output_obj
+        detached_output_obj = output_obj.clone()
```

Comment:
`clone()` will create a new tensor in graph. Releasing the copy will not affect origin output.

Reply:
> `clone()` will create a new tensor in graph. Releasing the copy will not affect origin output.

fixed in https://github.com/hpcaitech/ColossalAI/pull/6034/commits/b4103f125c0629e99cede00fef3ec5c67e6de74d;

Reply:
You'd better add `detach()` to avoid unexpected reference of graph